Modules/posixmodule.c

```diff
@@ -5490,12 +5490,101 @@ os__path_normpath_impl(PyObject *module, PyObject *path)
         return NULL;
     }
     Py_ssize_t norm_len;
-    wchar_t *norm_path = _Py_normpath_and_size(buffer, len, &norm_len);
+    wchar_t *norm_path = _Py_normpath_and_size(buffer, len, 0, &norm_len);
     PyObject *result = PyUnicode_FromWideChar(norm_path, norm_len);
     PyMem_Free(buffer);
     return result;
 }
 
+#ifndef MS_WINDOWS
+
+/*[clinic input]
+os._path_abspath
+
+    path: unicode
+    /
+
+Make path absolute.
+[clinic start generated code]*/
+
+static PyObject *
+os__path_abspath_impl(PyObject *module, PyObject *path)
+/*[clinic end generated code: output=b58956d662b60be0 input=577ecb3473d22113]*/
+{
+    Py_ssize_t path_len, start, abs_len;
+    wchar_t *abs, *abs_buf = NULL;
+    PyObject *result = NULL;
+
+    wchar_t *path_buf = PyUnicode_AsWideCharString(path, &path_len);
+    if (!path_buf) {
+        goto exit;
+    }
+
+    if (path_len == 0 || (path_len == 1 && path_buf[0] == L'.')) {
+        result = posix_getcwd(0);
+        goto exit;
+    }
+
+    if (_Py_isabs(path_buf)) {
+        start = 0;
+        abs_len = path_len;
+        abs = path_buf;
+    }
+    else {
+        PyObject *cwd_obj = posix_getcwd(0);
+        if (!cwd_obj) {
+            goto exit;
+        }
+        Py_ssize_t cwd_len;
+        wchar_t *cwd_buf = PyUnicode_AsWideCharString(cwd_obj, &cwd_len);
+        Py_DECREF(cwd_obj);
+        if (!cwd_buf) {
+            goto exit;
+        }
+
+        int add_sep = cwd_buf[cwd_len - 1] != SEP;
+        start = cwd_len + add_sep;
+        abs_len = start + path_len;
+
+        if ((size_t)abs_len + 1 > (size_t)PY_SSIZE_T_MAX / sizeof(wchar_t)) {
+            PyErr_SetString(PyExc_OverflowError, "path is too long");
+            goto exit;
+        }
+
+        abs_buf = PyMem_RawMalloc(((size_t)abs_len + 1) * sizeof(wchar_t));
+        if (!abs_buf) {
+            PyErr_NoMemory();
+            goto exit;
+        }
+        abs = abs_buf;
+
+        // Join cwd & path
+        wchar_t *p = memcpy(abs_buf, cwd_buf, cwd_len * sizeof(wchar_t));
+        p += cwd_len;
+        if (add_sep) {
+            *(p++) = SEP;
+        }
+        memcpy(p, path_buf, path_len * sizeof(wchar_t));
+        p[path_len] = '\0';
+
+        PyMem_Free(cwd_buf);
```

Comment:
After browsing through this PR and Modules/posixmodule.c, I wonder why there is no file scoped C helper for joining paths already. There seems to be, but only for Windows. This can't be the only place in posixmodule.c where a POSIX helper needs to join paths.

Reply:
There is, but for bytes:

https://github.com/python/cpython/blob/5a0209fc23de113747058858a4d2e5fc8213711e/Modules/posixmodule.c#L15846-L15876